core/vm/validate.go

```diff
@@ -0,0 +1,229 @@
+	// Therefore, the last instruction must be considered terminal.
+	if !jt[op].terminal {
+		return fmt.Errorf("%w: end with %s, pos %d", ErrInvalidCodeTermination, op, i)
+	}
+	if paths, err := validateControlFlow(code, section, metadata, jt); err != nil {
+		return err
+	} else if paths != count {
+		// TODO(matt): return actual position of unreacable code
+		return ErrUnreachableCode
+	}
+	return nil
+}
+
+// checkDest parses a relative offset at code[0:2] and checks if it is a valid jump destination.
+func checkDest(code []byte, analysis *bitvec, imm, from, length int) error {
+	if len(code) < imm+2 {
+		return io.ErrUnexpectedEOF
+	}
+	if analysis != nil && *analysis == nil {
+		*analysis = eofCodeBitmap(code)
+	}
+	offset := parseInt16(code[imm:])
+	dest := from + offset
+	if dest < 0 || dest >= length {
+		return fmt.Errorf("%w: out-of-bounds offset: offset %d, dest %d, pos %d", ErrInvalidJumpDest, offset, dest, imm)
+	}
+	if !analysis.codeSegment(uint64(dest)) {
+		return fmt.Errorf("%w: offset into immediate: offset %d, dest %d, pos %d", ErrInvalidJumpDest, offset, dest, imm)
+	}
+	return nil
+}
+
+// validateControlFlow iterates through all possible branches the provided code
+// value and determines if it is valid per EOF v1.
+func validateControlFlow(code []byte, section int, metadata []*FunctionMetadata, jt *JumpTable) (int, error) {
+	type item struct {
+		pos    int
+		height int
+	}
+	var (
+		heights        = make(map[int]int)
+		worklist       = []item{{0, int(metadata[section].Input)}}
+		maxStackHeight = int(metadata[section].Input)
+	)
+	for 0 < len(worklist) {
+		var (
+			idx    = len(worklist) - 1
+			pos    = worklist[idx].pos
+			height = worklist[idx].height
+		)
+		worklist = worklist[:idx]
+	outer:
+		for pos < len(code) {
+			op := OpCode(code[pos])
+
+			// Check if pos has already be visited; if so, the stack heights should be the same.
+			if want, ok := heights[pos]; ok {
+				if height != want {
+					return 0, fmt.Errorf("%w: have %d, want %d", ErrConflictingStack, height, want)
+				}
+				// Already visited this path and stack height
+				// matches.
+				break
+			}
+			heights[pos] = height
+
+			// Validate height for current op and update as needed.
+			if want, have := jt[op].minStack, height; want > have {
+				return 0, fmt.Errorf("%w: at pos %d", ErrStackUnderflow{stackLen: have, required: want}, pos)
+			}
+			if want, have := jt[op].maxStack, height; want < have {
+				return 0, fmt.Errorf("%w: at pos %d", ErrStackOverflow{stackLen: have, limit: want}, pos)
+			}
+			height += int(params.StackLimit) - jt[op].maxStack
+
+			switch {
+			case op == CALLF:
+				arg, _ := parseUint16(code[pos+1:])
+				if want, have := int(metadata[arg].Input), height; want > have {
+					return 0, fmt.Errorf("%w: at pos %d", ErrStackUnderflow{stackLen: have, required: want}, pos)
+				}
+				if have, limit := int(metadata[arg].Output)+height, int(params.StackLimit); have > limit {
```

Comment:
I think this should check 
```suggestion
				if have, limit := int(metadata[arg].MaxStackHeight)+height, int(params.StackLimit); have > limit {
```